docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,186 @@
+code (which obviously comes with its own type system), the foreign bindings need to know
+lots of details about all the types expressed by the tool.
+
+For the sake of this document, we will use the term "type universe" to define the set of
+all types known by each of the tools. Both of these tools build their own "type universe" then
+use that to generate both Rust code and foreign bindings.
+
+## UniFFI's type universe
+UniFFI's model is to parse an external ffi description from a `.udl` file which describes the
+entire "type universe". This type universe is then used to generate both the Rust scaffolding
+(on disk as a `.rs` file) and the foreign bindings.
+
+**What's good about this** is that the entire type system is known when generating both the rust code
+and the foreign binding.
+
+**What's bad about this** is that the external UDL is very ugly and redundant in terms of the
+implemented rust API.
+
+## Diplomat's type universe
+
+Diplomat defines its "type universe" (ie, the external ffi) using macros.
+
+**What's good about this** is that the "ffi module" defines the canonical API and it is defined in
+terms of Rust types - the redundant UDL is removed. The Rust scaffolding can also be generated
+by the macros, meaning there are no generated `.rs` files involved.
+
+Ryan even tried this for UniFFI in [#416](https://github.com/mozilla/uniffi-rs/pull/416) - but we
+struck **what's bad about this**: the context in which the macro runs doesn't know about types defined
+outside of that macro, which are what we need to expose.
+
+## Limitations in the macro approach
+
+Let's look at diplomat's simple example:
+
+```rust
+#[diplomat::bridge]
+mod ffi {
+    pub struct MyFFIType {
+        pub a: i32,
+        pub b: bool,
+    }
+
+    impl MyFFIType {
+        pub fn create() -> MyFFIType { ... }
+        ...
+    }
+}
+```
+
+This works fine, but starts to come unstuck if you want the types defined somewhere else. In this trivial example, something like:
+
+```Rust
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+
+#[diplomat::bridge]
+mod ffi {
+    impl MyFFIType {
+        pub fn create() -> MyFFIType { ... }
+        ...
+    }
+}
+```
+
+fails - diplomat can't handle this scenario - in the same way and for the same reasons that Ryan's
+[#416](https://github.com/mozilla/uniffi-rs/pull/416) can't - the contents of the struct aren't known.
+
+From the Rust side of the world, this is probably solvable by sprinkling more macros around - eg, something like:
+
+```Rust
+#[uniffi::magic]
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+```
+
+Might be enough for the generation of the Rust scaffolding. However, the problems are in the foreign bindings.
+
+## How the type universe is constructed for the macro approach.
+
+In both diplomat and [#416](https://github.com/mozilla/uniffi-rs/pull/416), the approach taken
+is that the generation process wants a path to the Rust source file that contains the module in
```

Comment:
> there may be multiple #[diplomat::bridge] tagged modules in the crate

@mhammond @badboy I wonder how much of a difference this would make to a use-case like Glean, if you could e.g. have one `#[diplomat::bridge]` declaration per metric type but the tool knows how to find them all and expose them as a single API surface when generating the bindings.

Reply:
This is certainly worth experimenting with, especially if we can avoid the "module" limitation - eg, if all types needed `[uniffi::magic]` then maybe there's actually no reason to insist on a wrapping module? Regardless, I've tried to capture this idea in the doc.

Reply:
I think in the 416 approach that would still have not worked, because we hash all input to put that into the generated function names as a way to prevent use of the wrong version of a library.
A single `#[diplomat::bridge]` invocation cannot know all input and thus can't generate one hash.
Of course this is a limitation we built ourselves, so maybe could be lifted.

Reply:
Re-reading quickly how diplomat works that same hashing could be applied to only the limited bridged module while still giving the same benefits.
Given that `diplomat-tool` parses the same code, it would come up with the same hash and know what functions to call in the foreign language code.

Reply:
Yeah the reason we have the module wrapper is mostly for convenience, because we need to look at imports as well sometimes.

Reply:
Yeah, thanks for clarifying - and while that's interesting context for this discussion, I don't think a discussion of the hashing needs to go into that document. LMK if you disagree though.

Reply:
Leaving that out sounds right to me.
It's a thing we will need to consider if we try the macro approach again, but not necessary to discuss in comparison/

Reply:
> Re-reading quickly how diplomat works that same hashing could be applied to only the limited bridged module
> while still giving the same benefits.

Good point, and in retrospect this thing about including a hash of the component in the name of the generated function probably deserved its own separate design doc. A summary of what's in my head based on the above, in case some of the details aren't obvious:

The main goal of naming our FFI functions like `component_<hash>_do_the_thing` instead of just `component_do_the_thing` is to guard against undefined behaviours if foreign-language bindings for version X of the interface are accidentally used with a `.so` for version Y of the crate. That is: if the details of how to call one of the FFI functions change, then we want the generated name of that function to change.

We don't necessarily have to achieve that by using a single `<hash>` of the whole API surface. If we split the FFI into several independent macro invocations, then as long as they're *self-contained* then they could all safely use their own individual hash.